server/src/main/java/org/cloudfoundry/identity/uaa/scim/vmidentity/VmidentityScimGroupMembershipManager.java

```diff
@@ -0,0 +1,343 @@
+                    for (Group g : groups) {
+                        members.add(createGroupMembership(g.getId().getUPN(), ScimGroupMember.Type.GROUP,
+                                groupInSystemDomain));
+                    }
+                }
+
+                if (users != null) {
+                    for (PersonUser user : users) {
+                        members.add(createGroupMembership(user.getId().getUPN(), ScimGroupMember.Type.USER,
+                                groupInSystemDomain));
+                    }
+                }
+            }
+
+            return members;
+        } catch (InvalidPrincipalException ex) {
+            logger.debug(String.format("Scim resource not found '%s'.", ex.getPrincipal()));
+            throw new ScimResourceNotFoundException(String.format("Scim resource not found '%s'.", ex.getPrincipal()));
+        } catch (Exception ex) {
+            logger.debug(String.format("getMembers '%s', '%s' ", groupId, permission.name()), ex);
+            throw new IllegalStateException(String.format("getMembers '%s', '%s' ", groupId, permission.name()), ex);
+        }
+    }
+
+    @Override
+    public Set<ScimGroup> getGroupsWithMember(String memberId, boolean transitive)
+            throws ScimResourceNotFoundException {
+        try {
+            HashSet<ScimGroup> members = new HashSet<ScimGroup>();
+            String tenant = VmidentityUtils.getTenantName(this.idmClient.getSystemTenant());
+            String systemDomain = VmidentityUtils.getSystemDomain(tenant, this.idmClient);
+
+            PrincipalId principalId = VmidentityUtils.getPrincipalId(memberId);
+            Set<Group> groups = null;
+
+            if (transitive) {
+                groups = this.idmClient.findNestedParentGroups(tenant, principalId);
+            } else {
+                groups = this.idmClient.findDirectParentGroups(tenant, principalId);
+            }
+
+            if (groups != null) {
+                for (Group g : groups) {
+                    members.add(VmidentityScimGroupProvisioning.getScimGroup(g, tenant, systemDomain));
+                }
+            }
+
+            return members;
+        } catch (InvalidPrincipalException ex) {
+            logger.debug(String.format("Scim resource not found '%s'.", ex.getPrincipal()));
+            throw new ScimResourceNotFoundException(String.format("Scim resource not found '%s'.", ex.getPrincipal()));
+        } catch (Exception ex) {
+            logger.debug(String.format("getGroupsWithMember '%s', '%s' ", memberId, transitive ? "true" : "false"), ex);
+            throw new IllegalStateException(
+                    String.format("getGroupsWithMember '%s', '%s' ", memberId, transitive ? "true" : "false"), ex);
+        }
+    }
+
+    @Override
+    public ScimGroupMember getMemberById(String groupId, String memberId)
+            throws ScimResourceNotFoundException, MemberNotFoundException {
+        // todo: implement better perf
```

Comment:
Not as part of this task, since it isn't SCIM related. I can go through the TODOs and tack them into a general-purpose 'perf' task later though.